### ISSUE TITLE
Bug :Pytorch and ONNX directml issues

### ISSUE THREAD

**Issue by Jay19751103** [ultralytics/ultralytics#7767]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

Hardware use NVIDIA 3080TI or AMD 7900XTX
Software : pytorch dml or ONNX dml (modify autobackend.py to add DmlExecutionProvider from 
https://www.linkedin.com/posts/arian-rahmatpour-a45200284_deeplearning-ai-gpu-activity-7112846015090544641-BvoL)

Model yolov8x-pose.pt

For Pytorch directml
there is one OP not support ultralytics\utils\ops.py
    xc = prediction[:, 4:mi].amax(1) > conf_thres  # candidates
Is it possible change to old version ?
    xc = prediction[:, 4:mi].max(1).values > conf_thres

and time.time return is strange, it will return long postprocessing
Is it possible to change it to time.perf_counter() ?

For ONNX
time.time return is strange as pytorch directml

Refer to issue 
https://github.com/ultralytics/yolov5/issues/11767

Following is test results with several runs (related to ultralytics\utils\ops.py time function)
time.time() output
0: 640x480 4 persons, 0.0ms
Speed: 0.0ms preprocess, 0.0ms inference, 15.6ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 0.0ms
Speed: 0.0ms preprocess, 0.0ms inference, 15.6ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 0.0ms
Speed: 0.0ms preprocess, 0.0ms inference, 15.6ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 0.0ms
Speed: 0.0ms preprocess, 0.0ms inference, 15.6ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 0.0ms
Speed: 0.0ms preprocess, 0.0ms inference, 15.6ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 0.0ms
Speed: 0.0ms preprocess, 0.0ms inference, 17.7ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 10.0ms
Speed: 0.0ms preprocess, 10.0ms inference, 4.0ms postprocess per image at shape (1, 3, 640, 480)

time.perf_counter() output
0: 640x480 4 persons, 7.9ms
Speed: 1.9ms preprocess, 7.9ms inference, 4.3ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 7.9ms
Speed: 2.0ms preprocess, 7.9ms inference, 4.3ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 7.9ms
Speed: 2.1ms preprocess, 7.9ms inference, 4.3ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 6.9ms
Speed: 1.9ms preprocess, 6.9ms inference, 5.7ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 6.7ms
Speed: 2.0ms preprocess, 6.7ms inference, 5.5ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 8.1ms
Speed: 2.0ms preprocess, 8.1ms inference, 4.2ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 7.9ms
Speed: 2.1ms preprocess, 7.9ms inference, 4.3ms postprocess per image at shape (1, 3, 640, 480)

0: 640x480 4 persons, 6.9ms
Speed: 1.7ms preprocess, 6.9ms inference, 5.7ms postprocess per image at shape (1, 3, 640, 480)

### Environment

[requirements.txt](https://github.com/ultralytics/ultralytics/files/14021694/requirements.txt)
[update_for_pytorch.zip](https://github.com/ultralytics/ultralytics/files/14021720/update_for_pytorch.zip)
for pytorch directml 
conda create with python 3.10 enviroment
pip install -r requirements
replace two files, head.py and torch_utils.py
running main.py 

for onnx dml 
pip install -r requirements
pip uninstall pytorch-directml
pip install onnx onnxruntime onnxruntime-directml

use yolo export model=modelname format=onnx to export for onnx format

(yolov8) D:\DirectML\ultralytics\ultralytics>git diff
diff --git a/ultralytics/nn/autobackend.py b/ultralytics/nn/autobackend.py
index 8f55c3ba..2a08979f 100644
--- a/ultralytics/nn/autobackend.py
+++ b/ultralytics/nn/autobackend.py
@@ -174,7 +174,7 @@ class AutoBackend(nn.Module):
             check_requirements(("onnx", "onnxruntime-gpu" if cuda else "onnxruntime"))
             import onnxruntime

-            providers = ["CUDAExecutionProvider", "CPUExecutionProvider"] if cuda else ["CPUExecutionProvider"]
+            providers = ["CUDAExecutionProvider", "CPUExecutionProvider"] if cuda else ["DmlExecutionProvider"]
             session = onnxruntime.InferenceSession(w, providers=providers)
             output_names = [x.name for x in session.get_outputs()]
             metadata = session.get_modelmeta().custom_metadata_map  # metadata
diff --git a/ultralytics/utils/ops.py b/ultralytics/utils/ops.py
index 8a86239c..abf50a63 100644
--- a/ultralytics/utils/ops.py
+++ b/ultralytics/utils/ops.py
@@ -58,7 +58,8 @@ class Profile(contextlib.ContextDecorator):
         """Get current time."""
         if self.cuda:
             torch.cuda.synchronize()
-        return time.time()
+#        return time.time()
+        return time.perf_counter()

###############optional to let pytorch directml support op without warning message.
 def segment2box(segment, width=640, height=640):
@@ -214,7 +215,8 @@ def non_max_suppression(
     nc = nc or (prediction.shape[1] - 4)  # number of classes
     nm = prediction.shape[1] - nc - 4
     mi = 4 + nc  # mask start index
-    xc = prediction[:, 4:mi].amax(1) > conf_thres  # candidates
+#    xc = prediction[:, 4:mi].amax(1) > conf_thres  # candidates
+    xc = prediction[:, 4:mi].max(1).values > conf_thres

     # Settings
     # min_wh = 2  # (pixels) minimum box width and height

### Minimal Reproducible Example

from ultralytics import YOLO
import cv2
import time

model = YOLO("yolov8x-pose.pt")

im2 = cv2.imread("bus.jpg")
for i in range(10):
    print("Trail time %d" % i)
    s = time.time()
    results = model.predict(source=im2, device='dml')
    e = time.time()
    print("----", e-s)
print("----")

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Jay19751103 thank you for the detailed bug report. It seems you're encountering issues with PyTorch DirectML and ONNX DirectML backends, specifically with certain operations and timing functions.

For the PyTorch DirectML operation, we aim to maintain compatibility with the latest stable versions of PyTorch. Changing the operation as you suggested may have broader implications, so we would need to review it thoroughly to ensure it doesn't affect other parts of the codebase or performance on other platforms.

Regarding the timing function, `time.perf_counter()` is indeed a more precise timing function compared to `time.time()`, and we will consider updating this in future releases to provide more accurate timing measurements.

For the ONNX issue, it's similar to the timing function in PyTorch DirectML. We appreciate your suggestion and will take it into account for future updates.

If you're experiencing long postprocessing times, it could be related to the backend you're using. DirectML is less common and might not be as optimized as CUDA for certain operations.

We always welcome contributions from the community. If you feel confident in making these changes and can ensure they don't negatively impact other users, we encourage you to submit a PR for review. Your contributions help improve YOLOv8 for everyone. 🚀

For further guidance on contributing, please refer to our documentation on how to contribute to the project. Thank you for your support and for being part of the YOLO community!

**Comment by Jay19751103** [comment]
Hi @glenn-jocher
I try tensor RT with python also have same problem
Inference on 3080TI
tensor RT with C++ application take 4.x ms (batch 8 will be 2.x ms)
tensor RT with Python  after patch will be identical with C++ 4.xms

**Comment by glenn-jocher** [comment]
@Jay19751103 thank you for reporting this issue with detailed information and for your suggestions on potential fixes. Let's address each point you've raised:

1. **PyTorch DirectML OP Support**: The change you suggested from `amax(1)` to `max(1).values` is indeed a valid workaround for operations not supported by PyTorch DirectML. However, this change may affect performance on other backends. We will consider this change and test its impact across different environments before making a decision.

2. **Time Measurement**: The use of `time.time()` vs. `time.perf_counter()` is a valid point, especially when high-resolution timers are needed. `time.perf_counter()` provides a higher resolution timer which is more suitable for measuring short durations. We will evaluate the impact of switching to `time.perf_counter()` in the context of the entire codebase and make the necessary adjustments.

3. **ONNX Export and Inference**: The strange timing results you're observing with `time.time()` could be due to the lower resolution of this timer function. Switching to `time.perf_counter()` as you suggested should provide more accurate timing measurements. We will investigate this further and apply the change if it proves to be beneficial.

4. **ONNX DirectML Issues**: The timing issues you've observed with ONNX DirectML are likely related to the timer function as well. We will investigate this and consider using `time.perf_counter()` for more accurate measurements.

For the PyTorch DirectML OP support, we can consider your suggestion as a temporary workaround for users facing similar issues. However, we need to ensure that any changes made do not negatively impact other users who are not using DirectML.

For the timing issues, we will investigate the use of `time.perf_counter()` and make changes accordingly if it provides more accurate and consistent timing measurements across different platforms and environments.

We appreciate your willingness to help and your detailed bug report. If you're interested in contributing to the resolution of this issue, please let us know, and we can guide you on how to submit a PR. Otherwise, we will take your report into consideration and work on resolving these issues in upcoming updates.
